pandas/_libs/index.pyx

```diff
@@ -1099,3 +1126,123 @@ cdef class ExtensionEngine(SharedEngine):
 
     cdef _check_type(self, object val):
         hash(val)
+
+
+cdef class MaskedIndexEngine(IndexEngine):
+
+    def __init__(self, ndarray values, ndarray mask):
+        super().__init__(values, mask)
+
+    def get_indexer_non_unique(self, ndarray targets, ndarray target_mask):
```

Comment:
no way to de-duplicate this?

Reply:
We could try to merge some parts of the function, but would rather do this as follow up to keep diffs simpler

Reply:
fair enough

Reply:
can add a TODO to that effect

Reply:
Added